utils/s2n_safety.h

```diff
@@ -111,3 +111,17 @@ extern int s2n_mul_overflow(uint32_t a, uint32_t b, uint32_t* out);
 extern int s2n_align_to(uint32_t initial, uint32_t alignment, uint32_t* out);
 extern int s2n_add_overflow(uint32_t a, uint32_t b, uint32_t* out);
 extern int s2n_sub_overflow(uint32_t a, uint32_t b, uint32_t* out);
+
+/**
+ * Define a C99-compliant static assert.
+ *
+ * Static_assert() is not defined before C11. So, we can't rely on it when we
+ * want to be C99 compliant. Instead, use a method that is guaranteed to be C99
+ * compliant and still give us an equivalent static assert mechanism.
+ *
+ * Taken from aws-c-common: https://github.com/awslabs/aws-c-common/blob/main/include/aws/common/macros.h#L19
+ */
+#define S2N_CONCAT(A, B) A##B
+#define S2N_STATIC_ASSERT0(cond, msg) typedef char S2N_CONCAT(static_assertion_, msg)[(!!(cond)) * 2 - 1]
+#define S2N_STATIC_ASSERT1(cond, line) S2N_STATIC_ASSERT0(cond, S2N_CONCAT(at_line_, line))
+#define S2N_STATIC_ASSERT(cond) S2N_STATIC_ASSERT1(cond, __LINE__)
```

Comment:
Do you have an example of what it should look like when a static assertion fails? I tried it in the compiler explorer, but the error didn't look particularly helpful: https://godbolt.org/z/b6joz4vMP

And do we need to split this macro up like this? I think it makes it harder to follow, not easier. Isn't it just:
```suggestion
#define S2N_STATIC_ASSERT(cond) \
    typedef char static_assertion_at_line_##__LINE__[(!!(cond)) * 2 - 1]
```

Reply:
The error looks a bit convoluted agreed. In AWS-LC we also tag it with a reason which gives context - but it is pretty much the best we can do AFAIK when we have to go down to C99 that doesn't have a build in static assert.

I don't think the latter would work, because `__LINE__` is not expanded.

Reply:
I don't see it expanded in the current version either. Do you have an example of the expected output?

Reply:
The error you get with your godbolt reference is
```
<source>:2:102: error: size '-1' of array 'static_assertion_at_line_6' is negative
```
There is tracing in the output, so it is not straightforward indeed. But I don't know of any other methods that can give me a static assertion.

 I just use this to fail the build if my assumption is untrue and I really want it here: I use it to verify the assumption `MADV_WIPEONFORK = 18`, if the macro is defined.

Reply:
Oh I think I see. The `__LINE__` appended is about making the array used for the assertion unique, not to try to help with debugging. This is pretty hacky.

Any reason you don't just use the #error directive? Like
```
#if <condition>
#error "oh no"
#endif
```